src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs

```diff
@@ -0,0 +1,1280 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    [CompilerTrait(CompilerFeature.StackAllocInitializer)]
+    public class StackAllocInitializerTests : CompilingTestBase
+    {
+        [Fact]
+        public void NoBestType()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+unsafe class Test
+{
+    struct A {}
+    struct B {}
+
+    public void Method1()
+    {
+        var obj1 = stackalloc[] { new A(), new B() };
+    }
+}", TestOptions.UnsafeReleaseDll);
+
+            comp.VerifyDiagnostics(
+                // (9,20): error CS0826: No best type found for implicitly-typed array
+                //         var obj1 = stackalloc[] { new A(), new B() };
+                Diagnostic(ErrorCode.ERR_ImplicitlyTypedArrayNoBestType, "stackalloc[] { new A(), new B() }").WithLocation(9, 20)
+                );
+        }
+
+        [Fact]
+        public void BestTypeNumeric()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+unsafe class Test
+{
+    public void Method1()
+    {
+        var obj1 = stackalloc[] { 1, 1.2 };
+    }
+}", TestOptions.UnsafeReleaseDll);
+
+            comp.VerifyDiagnostics(
+                );
+
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree);
+
+            var variables = tree.GetCompilationUnitRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>();
+            Assert.Equal(1, variables.Count());
+
+            var obj1 = variables.ElementAt(0);
+            Assert.Equal("obj1", obj1.Identifier.Text);
+
+            var obj1Value = model.GetSemanticInfoSummary(obj1.Initializer.Value);
+            Assert.Equal(SpecialType.System_Double, ((PointerTypeSymbol)obj1Value.Type).PointedAtType.SpecialType);
+        }
```

Comment:
Let's test other semantic model APIs too. There is a list at [https://github.com/dotnet/roslyn/blob/master/docs/contributing/Compiler%20Test%20Plan.md](https://github.com/dotnet/roslyn/blob/master/docs/contributing/Compiler%2520Test%2520Plan.md)
In particular, `GetTypeInfo` seems relevant.
I'd also check `GetDeclaredSymbol` (should return nothing).

Reply:
GetSemanticInfoSummary calls GetTypeInfo, I'll will assert relevant properties.

Reply:
Sorry I wasn't clear. I mean testing `GetTypeInfo` directly on the syntax of the stackalloc.
```C#
var syntax = tree.GetRoot().DescendantNodes().OfType<Stackalloc...>().Single();
var typeInfo = model.GetTypeInfo(syntax); // expect Type and ConvertedType as appropriate (Span and pointer scenarios)
var declared = model.GetDeclaredSymbol(syntax); // expect nothing
```

Please also test this when the type isn't bound (as discussed in `BindStackAllocArrayCreationExpression`), so that we know the behavior (ie. hopefully not crashing).

---
In reply to: [166793581](https://github.com/dotnet/roslyn/pull/24249#discussion_r166793581) [](ancestors = 166793581)